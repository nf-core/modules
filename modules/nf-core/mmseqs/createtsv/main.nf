
process MMSEQS_CREATETSV {
    tag "$meta.id"
    label 'process_single'

    conda "bioconda::mmseqs2=14.7e284"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/mmseqs2:14.7e284--pl5321h6a68c12_2':
        'biocontainers/mmseqs2:14.7e284--pl5321h6a68c12_2' }"

    input:
    tuple val(meta), path(db_result)
    tuple val(meta2), path(db_query)
    tuple val(meta3), path(db_target)

    output:
    tuple val(meta), path("*.tsv"), emit: tsv
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args ?: "*.dbtype"
    def args3 = task.ext.args ?: "*.dbtype"
    def args4 = task.ext.args ?: "*.dbtype"
    def prefix = task.ext.prefix ?: "${meta.id}"
    db_target = db_target ?: "${db_query}" // optional argument db_target as in many cases, it's the same as db_query
    """
    # Extract files with specified args based suffix | remove suffix | isolate longest common substring of files
    DB_RESULT_PATH_NAME=\$(find -L "$db_result/" -maxdepth 1 -name "$args2" | sed 's/\\.\\[^.\\]*\$//' | sed -e 'N;s/^\\(.*\\).*\\n\\1.*\$/\\1\\n\\1/;D' )
    DB_QUERY_PATH_NAME=\$(find -L "$db_query/" -maxdepth 1 -name "$args3" | sed 's/\\.\\[^.\\]*\$//' | sed -e 'N;s/^\\(.*\\).*\\n\\1.*\$/\\1\\n\\1/;D' )
    DB_TARGET_PATH_NAME=\$(find -L "$db_target/" -maxdepth 1 -name "$args4" | sed 's/\\.\\[^.\\]*\$//' | sed -e 'N;s/^\\(.*\\).*\\n\\1.*\$/\\1\\n\\1/;D' )

    mmseqs \\
        createtsv \\
        \$DB_QUERY_PATH_NAME \\
        \$DB_TARGET_PATH_NAME \\
        \$DB_RESULT_PATH_NAME \\
        ${prefix}.tsv \\
        $args \\
        --threads ${task.cpus} \\
        --compressed 1

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        mmseqs: \$(mmseqs | grep 'Version' | sed 's/MMseqs2 Version: //')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        mmseqs: \$(mmseqs | grep 'Version' | sed 's/MMseqs2 Version: //')
    END_VERSIONS
    """
}
