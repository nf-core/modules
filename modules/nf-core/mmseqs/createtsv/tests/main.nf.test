nextflow_process {

    name "Test Process MMSEQS_CREATETSV"
    script "../main.nf"
    process "MMSEQS_CREATETSV"

    tag "modules"
    tag "modules_nfcore"
    tag "mmseqs"
    tag "mmseqs/taxonomy"
    tag "mmseqs/createdb"
    tag "mmseqs/databases"
    tag "untar"
    tag "mmseqs/createtsv"

    test("mmseqs/createtsv - bacteroides_fragilis - taxonomy") {

        config  "./taxonomy.nextflow.config"

        setup {
            run("MMSEQS_CREATEDB", alias: "MMSEQS_TAXA") {
                script "../../createdb/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test_query', single_end:false ],
                        file(params.test_data['bacteroides_fragilis']['genome']['genome_fna_gz'], checkIfExists: true)
                        ]
                    """
                }
            }
            run("MMSEQS_DATABASES") {
                script "../../databases/main.nf"
                process {
                """
                input[0] = 'SILVA'
                """
                }
            }
            run("MMSEQS_TAXONOMY") {
                script "../../taxonomy/main.nf"
                process {
                """
                input[0] = MMSEQS_TAXA.out.db
                input[1] = MMSEQS_DATABASES.out.database
                """
                }
            }
        }
        when {
            process {
                """
                input[0] = MMSEQS_TAXONOMY.out.db_taxonomy
                input[1] = [[:],[]]
                input[2] = MMSEQS_TAXA.out.db
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("mmseqs/createtsv - sarscov2 - cluster") {

        config  "./cluster.nextflow.config"

        setup {
            run("UNTAR", alias: "UNTAR_QUERY") {
                script "../../../untar/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test_query', single_end:true ],
                        file(params.test_data['sarscov2']['genome']['mmseqs_tar_gz'], checkIfExists: true),
                    ]
                    """
                }
            }
            run("UNTAR", alias: "UNTAR_TARGET") {
                script "../../../untar/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test_target', single_end:true ],
                        file(params.test_data['sarscov2']['genome']['mmseqs_tar_gz'], checkIfExists: true),
                    ]
                    """
                }
            }
            run("UNTAR", alias: "UNTAR_RESULT") {
                script "../../../untar/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test_result', single_end:true ],
                        file(params.test_data['sarscov2']['genome']['mmseqs_tar_gz'], checkIfExists: true),
                    ]
                    """
                }
            }
        }

        when {

            process {
                """
                ch_query  = UNTAR_QUERY.out.untar
                ch_target = UNTAR_TARGET.out.untar
                ch_result = UNTAR_RESULT.out.untar

                input[0] = ch_result
                input[1] = ch_query
                input[2] = ch_target
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.tsv).match("tsv") },
                { assert snapshot(process.out.versions).match("versions") }
            )
        }
    }

    test("mmseqs/createtsv - bacteroides_fragilis - taxonomy - stub") {

        options "-stub"
        config  "./taxonomy.nextflow.config"

        setup {
            run("MMSEQS_CREATEDB", alias: "MMSEQS_TAXA") {
                script "../../createdb/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test_query', single_end:false ],
                        file(params.test_data['bacteroides_fragilis']['genome']['genome_fna_gz'], checkIfExists: true)
                        ]
                    """
                }
            }
            run("MMSEQS_DATABASES") {
                script "../../databases/main.nf"
                process {
                """
                input[0] = 'SILVA'
                """
                }
            }
            run("MMSEQS_TAXONOMY") {
                script "../../taxonomy/main.nf"
                process {
                """
                input[0] = MMSEQS_TAXA.out.db
                input[1] = MMSEQS_DATABASES.out.database
                """
                }
            }
        }
        when {
            process {
                """
                input[0] = MMSEQS_TAXONOMY.out.db_taxonomy
                input[1] = [[:],[]]
                input[2] = MMSEQS_TAXA.out.db
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("mmseqs/createtsv - sarscov2 - cluster - stub") {

        options "-stub"
        config  "./cluster.nextflow.config"

        setup {
            run("UNTAR", alias: "UNTAR_QUERY") {
                script "../../../untar/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test_query', single_end:true ],
                        file(params.test_data['sarscov2']['genome']['mmseqs_tar_gz'], checkIfExists: true),
                    ]
                    """
                }
            }
            run("UNTAR", alias: "UNTAR_TARGET") {
                script "../../../untar/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test_target', single_end:true ],
                        file(params.test_data['sarscov2']['genome']['mmseqs_tar_gz'], checkIfExists: true),
                    ]
                    """
                }
            }
            run("UNTAR", alias: "UNTAR_RESULT") {
                script "../../../untar/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test_result', single_end:true ],
                        file(params.test_data['sarscov2']['genome']['mmseqs_tar_gz'], checkIfExists: true),
                    ]
                    """
                }
            }
        }

        when {

            process {
                """
                ch_query  = UNTAR_QUERY.out.untar
                ch_target = UNTAR_TARGET.out.untar
                ch_result = UNTAR_RESULT.out.untar

                input[0] = ch_result
                input[1] = ch_query
                input[2] = ch_target
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}