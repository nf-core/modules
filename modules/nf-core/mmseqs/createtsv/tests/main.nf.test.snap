{
    "versions": {
        "content": [
            [
                "versions.yml:md5,20a853f50c920d431e5ab7593ca79e6f"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-12T11:53:02.392516336"
    },
    "tsv": {
        "content": [
            [
                [
                    {
                        "id": "test_result",
                        "single_end": true
                    },
                    "test_result.tsv:md5,4e7ba50ce2879660dc6595286bf0d097"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-12T11:53:02.311022721"
    },
    "mmseqs/createtsv - bacteroides_fragilis - taxonomy - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_query",
                            "single_end": false
                        },
                        "test_query.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,20a853f50c920d431e5ab7593ca79e6f"
                ],
                "tsv": [
                    [
                        {
                            "id": "test_query",
                            "single_end": false
                        },
                        "test_query.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,20a853f50c920d431e5ab7593ca79e6f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-12T11:53:11.715695614"
    },
    "mmseqs/createtsv - sarscov2 - cluster - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_result",
                            "single_end": true
                        },
                        "test_result.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,20a853f50c920d431e5ab7593ca79e6f"
                ],
                "tsv": [
                    [
                        {
                            "id": "test_result",
                            "single_end": true
                        },
                        "test_result.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,20a853f50c920d431e5ab7593ca79e6f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-12T11:53:21.386186911"
    },
    "mmseqs/createtsv - bacteroides_fragilis - taxonomy": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_query",
                            "single_end": false
                        },
                        "test_query.tsv:md5,9179f5c85b8b87a4dc998c9d17840161"
                    ]
                ],
                "1": [
                    "versions.yml:md5,20a853f50c920d431e5ab7593ca79e6f"
                ],
                "tsv": [
                    [
                        {
                            "id": "test_query",
                            "single_end": false
                        },
                        "test_query.tsv:md5,9179f5c85b8b87a4dc998c9d17840161"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,20a853f50c920d431e5ab7593ca79e6f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-12T11:52:51.792345007"
    }
}
